Local Plugin  in 14.15.2 got Module not found: Error: Can't resolve

## Bug report

### Required System information
- Node.js version: 18.18.2
- NPM version:  8.19.1 
- Strapi version: 14.15.2 and 14.14.6
- Database: sqlite
- Operating system: window 11
- Is your project Javascript or Typescript: javascript

### Describe the bug

app.js  generated wrong path for local plugin in ./strapi/client 

### Steps to reproduce the behavior

1. Create plugins via cli
2. Update pluigin.js 
![image](https://github.com/strapi/strapi/assets/14286010/49333ccc-9848-4e85-9a51-7eb03fdcce1b)
3.. Run yarn build or strapi develop --watch-admin
5. Error
![image](https://github.com/strapi/strapi/assets/14286010/bc58218a-1cb2-4434-9208-23a4ed1e4013)
![image](https://github.com/strapi/strapi/assets/14286010/2c6d369b-3221-4d0f-856e-f37daf0e6341)
![image](https://github.com/strapi/strapi/assets/14286010/db90e80a-77da-4ee9-9cb5-ac60b5a56385)

### Expected behavior
Plugin should load correctly

Hi, sorry for late

> can you run `yarn build --debug` and share the entire log it produces? there will be a lot of `[DEBUG]: ` stuff that will help diagnose this i think...

Sorry for late response , this is my debug log. It happen only in 14.5.x , other version below ok.
`
$ strapi build --debug
[DEBUG] Loaded package.json: 
 {
  name: 'backend',
  private: true,
  version: '0.1.0',
  description: 'A Strapi application',
  scripts: {
    develop: 'strapi develop',
    start: 'strapi start',
    build: 'strapi build',
    strapi: 'strapi',
    postinstall: 'patch-package'
  },
  devDependencies: {},
  dependencies: {
    '@ckeditor/strapi-plugin-ckeditor': '^0.0.10',
    '@strapi/plugin-i18n': '4.15.2',
    '@strapi/plugin-seo': '^1.9.6',
    '@strapi/plugin-users-permissions': '4.15.2',
    '@strapi/provider-email-sendgrid': '4.15.2',
    '@strapi/strapi': '4.15.2',
    axios: '^1.4.0',
    'better-sqlite3': '^8.5.0',
    crypto: '^1.0.1',
    'ioredis-lock': '^4.0.0',
    md5: '^2.3.0',
    moment: '^2.29.4',
    'moment-timezone': '^0.5.43',
    mysql: '^2.18.1',
    'patch-package': '^7.0.2',
    qs: '^6.11.2',
    sha256: '^0.2.0',
    'strapi-advanced-uuid': '^1.0.11',
    'strapi-plugin-email-designer': '^2.2.0',
    'strapi-plugin-multi-select': '^1.2.2',
    'strapi-plugin-navigation': '^2.2.16',
    'strapi-plugin-react-icons': '^0.2.1',
    'strapi-plugin-redis': '^1.0.1',
    'strapi-plugin-rest-cache': '^4.2.8',
    'strapi-plugin-sitemap': '^2.0.11',
    'strapi-provider-rest-cache-redis': '^4.2.8'
  },
  author: { name: 'A Strapi developer' },
  strapi: { uuid: '0a730a29-e688-43a9-9793-8f1a7f696856' },
  engines: { node: '>=14.19.1 <=18.x.x', npm: '>=6.0.0' },
  license: 'MIT',
  readme: 'ERROR: No README data found!',
  _id: 'backend@0.1.0'
}
[INFO] The Strapi admin needs to install the following dependencies:
   - react@^18.0.0
  - react-dom@^18.0.0
  - react-router-dom@^5.2.0
  - styled-components@^5.2.1
? Would you like to install these dependencies now? These are not required but are recommended, from V5 these will be required. Yes
[INFO] Running 'yarn add react@^18.0.0 react-dom@^18.0.0 react-router-dom@^5.2.0 styled-components@^5.2.1'
[1/5] Validating package.json...
[2/5] Resolving packages...
[3/5] Fetching packages...
warning "@strapi/plugin-i18n > @strapi/design-system > @uiw/react-codemirror > @uiw/codemirror-extensions-basic-setup@4.21.20" has unmet peer dependency "@codemirror/view@>=6.0.0".
warning "@strapi/plugin-i18n > @strapi/design-system > @uiw/react-codemirror > @uiw/codemirror-extensions-basic-setup > @codemirror/autocomplete@6.11.0" has unmet peer dependency "@lezer/common@^1.0.0".
warning "@strapi/plugin-seo > eslint-plugin-react-hooks@4.6.0" has unmet peer dependency "eslint@^3.0.0 || ^4.0.0 || ^5.0.0 || ^6.0.0 || ^7.0.0 || ^8.0.0-0".
warning " > @strapi/plugin-users-permissions@4.15.2" has incorrect peer dependency "@strapi/strapi@4.0.0".
warning "@strapi/strapi > @strapi/admin@4.15.2" has incorrect peer dependency "@strapi/data-transfer@4.15.1".
warning "@strapi/strapi > @strapi/admin > styled-components > babel-plugin-styled-components > @babel/plugin-syntax-jsx@7.23.3" has unmet peer dependency "@babel/core@^7.0.0-0".
warning " > ioredis-lock@4.0.0" has unmet peer dependency "ioredis@~4.x.x".
warning " > styled-components@5.3.11" has unmet peer dependency "react-is@>= 16.8.0".
[5/5] Building fresh packages...

success Saved lockfile.
success Saved 1 new dependency.
info Direct dependencies
└─ styled-components@5.3.11
info All dependencies
└─ styled-components@5.3.11
$ patch-package
patch-package 7.0.2
Applying patches...
react-icons@4.8.0 ✔
strapi-plugin-email-designer@2.2.0 ✔
[DEBUG] Loaded package.json: 
 {
  name: 'backend',
  private: true,
  version: '0.1.0',
  description: 'A Strapi application',
  scripts: {
    develop: 'strapi develop',
    start: 'strapi start',
    build: 'strapi build',
    strapi: 'strapi',
    postinstall: 'patch-package'
  },
  devDependencies: {},
  dependencies: {
    '@ckeditor/strapi-plugin-ckeditor': '^0.0.10',
    '@strapi/plugin-i18n': '4.15.2',
    '@strapi/plugin-seo': '^1.9.6',
    '@strapi/plugin-users-permissions': '4.15.2',
    '@strapi/provider-email-sendgrid': '4.15.2',
    '@strapi/strapi': '4.15.2',
    axios: '^1.4.0',
    'better-sqlite3': '^8.5.0',
    crypto: '^1.0.1',
    'ioredis-lock': '^4.0.0',
    md5: '^2.3.0',
    moment: '^2.29.4',
    'moment-timezone': '^0.5.43',
    mysql: '^2.18.1',
    'patch-package': '^7.0.2',
    qs: '^6.11.2',
    react: '^18.0.0',
    'react-dom': '^18.0.0',
    'react-router-dom': '^5.2.0',
    sha256: '^0.2.0',
    'strapi-advanced-uuid': '^1.0.11',
    'strapi-plugin-email-designer': '^2.2.0',
    'strapi-plugin-multi-select': '^1.2.2',
    'strapi-plugin-navigation': '^2.2.16',
    'strapi-plugin-react-icons': '^0.2.1',
    'strapi-plugin-redis': '^1.0.1',
    'strapi-plugin-rest-cache': '^4.2.8',
    'strapi-plugin-sitemap': '^2.0.11',
    'strapi-provider-rest-cache-redis': '^4.2.8',
    'styled-components': '^5.2.1'
  },
  author: { name: 'A Strapi developer' },
  strapi: { uuid: '0a730a29-e688-43a9-9793-8f1a7f696856' },
  engines: { node: '>=14.19.1 <=18.x.x', npm: '>=6.0.0' },
  license: 'MIT',
  readme: 'ERROR: No README data found!',
  _id: 'backend@0.1.0'
}
- Building build context
[INFO] Including the following ENV variables as part of the JS bundle:
    - ADMIN_PATH
    - STRAPI_ADMIN_BACKEND_URL
    - STRAPI_TELEMETRY_DISABLED
[DEBUG] Cleaning dist folder: E:\projects\glamping\backend\build
[DEBUG] Cleaned dist folder
[DEBUG] Dependencies from user's project
 {
  '@ckeditor/strapi-plugin-ckeditor': '^0.0.10',
  '@strapi/plugin-i18n': '4.15.2',
  '@strapi/plugin-seo': '^1.9.6',
  '@strapi/plugin-users-permissions': '4.15.2',
  '@strapi/provider-email-sendgrid': '4.15.2',
  '@strapi/strapi': '4.15.2',
  axios: '^1.4.0',
  'better-sqlite3': '^8.5.0',
  crypto: '^1.0.1',
  'ioredis-lock': '^4.0.0',
  md5: '^2.3.0',
  moment: '^2.29.4',
  'moment-timezone': '^0.5.43',
  mysql: '^2.18.1',
  'patch-package': '^7.0.2',
  qs: '^6.11.2',
  react: '^18.0.0',
  'react-dom': '^18.0.0',
  'react-router-dom': '^5.2.0',
  sha256: '^0.2.0',
  'strapi-advanced-uuid': '^1.0.11',
  'strapi-plugin-email-designer': '^2.2.0',
  'strapi-plugin-multi-select': '^1.2.2',
  'strapi-plugin-navigation': '^2.2.16',
  'strapi-plugin-react-icons': '^0.2.1',
  'strapi-plugin-redis': '^1.0.1',
  'strapi-plugin-rest-cache': '^4.2.8',
  'strapi-plugin-sitemap': '^2.0.11',
  'strapi-provider-rest-cache-redis': '^4.2.8',
  'styled-components': '^5.2.1'
}
[DEBUG] User's plugins file 
 {
  'payment-gateway': { enabled: true, resolve: './src/plugins/payment-gateway' },
  redis: { config: { connections: [Object] } },
  'rest-cache': { config: { provider: [Object], strategy: [Object] } },
  email: {
    config: {
      provider: 'sendgrid',
      providerOptions: [Object],
      settings: [Object]
    }
  }
}
[DEBUG] Enabled plugins
 {
  ckeditor: {
    name: 'ckeditor',
    pathToPlugin: '@ckeditor/strapi-plugin-ckeditor'
  },
  i18n: { name: 'i18n', pathToPlugin: '@strapi/plugin-i18n' },
  seo: { name: 'seo', pathToPlugin: '@strapi/plugin-seo' },
  'users-permissions': {
    name: 'users-permissions',
    pathToPlugin: '@strapi/plugin-users-permissions'
  },
  'strapi-advanced-uuid': {
    name: 'strapi-advanced-uuid',
    pathToPlugin: 'strapi-advanced-uuid'
  },
  'email-designer': {
    name: 'email-designer',
    pathToPlugin: 'strapi-plugin-email-designer'
  },
  'multi-select': { name: 'multi-select', pathToPlugin: 'strapi-plugin-multi-select' },
  navigation: { name: 'navigation', pathToPlugin: 'strapi-plugin-navigation' },
  'react-icons': { name: 'react-icons', pathToPlugin: 'strapi-plugin-react-icons' },
  redis: { name: 'redis', pathToPlugin: 'strapi-plugin-redis' },
  'rest-cache': { name: 'rest-cache', pathToPlugin: 'strapi-plugin-rest-cache' },
  sitemap: { name: 'sitemap', pathToPlugin: 'strapi-plugin-sitemap' },
  'strapi-provider-rest-cache-redis': {
    name: 'strapi-provider-rest-cache-redis',
    pathToPlugin: 'strapi-provider-rest-cache-redis'
  },
  'payment-gateway': {
    name: 'payment-gateway',
    isLocal: true,
    pathToPlugin: './src/plugins/payment-gateway'
  }
}
[DEBUG] Enabled plugins with FE 
 {
  ckeditor: {
    name: 'ckeditor',
    pathToPlugin: '@ckeditor/strapi-plugin-ckeditor'
  },
  i18n: { name: 'i18n', pathToPlugin: '@strapi/plugin-i18n' },
  seo: { name: 'seo', pathToPlugin: '@strapi/plugin-seo' },
  'users-permissions': {
    name: 'users-permissions',
    pathToPlugin: '@strapi/plugin-users-permissions'
  },
  'strapi-advanced-uuid': {
    name: 'strapi-advanced-uuid',
    pathToPlugin: 'strapi-advanced-uuid'
  },
  'email-designer': {
    name: 'email-designer',
    pathToPlugin: 'strapi-plugin-email-designer'
  },
  'multi-select': { name: 'multi-select', pathToPlugin: 'strapi-plugin-multi-select' },
  navigation: { name: 'navigation', pathToPlugin: 'strapi-plugin-navigation' },
  'react-icons': { name: 'react-icons', pathToPlugin: 'strapi-plugin-react-icons' },
  redis: { name: 'redis', pathToPlugin: 'strapi-plugin-redis' },
  'rest-cache': { name: 'rest-cache', pathToPlugin: 'strapi-plugin-rest-cache' },
  sitemap: { name: 'sitemap', pathToPlugin: 'strapi-plugin-sitemap' },
  'strapi-provider-rest-cache-redis': {
    name: 'strapi-provider-rest-cache-redis',
    pathToPlugin: 'strapi-provider-rest-cache-redis'
  },
  'payment-gateway': {
    name: 'payment-gateway',
    isLocal: true,
    pathToPlugin: './src/plugins/payment-gateway'
  }
}
√ Building build context (140.04739999771118ms)
- Building admin panel
[DEBUG] Created the runtime directory
[DEBUG] Rendered the HTML
\ Building admin panel[DEBUG] Wrote the index.html file
[DEBUG] Wrote the app.js file
| Building admin panel[DEBUG] Webpack config {
  experiments: { topLevelAwait: true },
  entry: { main: [ './.strapi\\client\\app.js' ] },
  resolve: {
    alias: {
      '@strapi/design-system$': 'E:\\projects\\glamping\\backend\\node_modules\\@strapi\\design-system',
      '@strapi/helper-plugin$': 'E:\\projects\\glamping\\backend\\node_modules\\@strapi\\helper-plugin',
      '@strapi/icons$': 'E:\\projects\\glamping\\backend\\node_modules\\@strapi\\icons',
      'date-fns$': 'E:\\projects\\glamping\\backend\\node_modules\\date-fns',
      'formik$': 'E:\\projects\\glamping\\backend\\node_modules\\formik',
      'history$': 'E:\\projects\\glamping\\backend\\node_modules\\history',
      'immer$': 'E:\\projects\\glamping\\backend\\node_modules\\immer',
      'qs$': 'E:\\projects\\glamping\\backend\\node_modules\\@strapi\\admin\\node_modules\\qs',
      'lodash$': 'E:\\projects\\glamping\\backend\\node_modules\\lodash',
      'react$': 'E:\\projects\\glamping\\backend\\node_modules\\react',
      'react-dnd$': 'E:\\projects\\glamping\\backend\\node_modules\\react-dnd',
      'react-dnd-html5-backend$': 'E:\\projects\\glamping\\backend\\node_modules\\react-dnd-html5-backend',
      'react-dom$': 'E:\\projects\\glamping\\backend\\node_modules\\react-dom',
      'react-error-boundary$': 'E:\\projects\\glamping\\backend\\node_modules\\react-error-boundary',
      'react-helmet$': 'E:\\projects\\glamping\\backend\\node_modules\\react-helmet',
      'react-is$': 'E:\\projects\\glamping\\backend\\node_modules\\@strapi\\admin\\node_modules\\react-is',
      'react-intl$': 'E:\\projects\\glamping\\backend\\node_modules\\react-intl',
      'react-query$': 'E:\\projects\\glamping\\backend\\node_modules\\react-query',
      'react-redux$': 'E:\\projects\\glamping\\backend\\node_modules\\react-redux',
      'react-router-dom$': 'E:\\projects\\glamping\\backend\\node_modules\\react-router-dom',
      'react-window$': 'E:\\projects\\glamping\\backend\\node_modules\\react-window',
      'react-select$': 'E:\\projects\\glamping\\backend\\node_modules\\react-select',
      'redux$': 'E:\\projects\\glamping\\backend\\node_modules\\redux',
      'reselect$': 'E:\\projects\\glamping\\backend\\node_modules\\reselect',
      'styled-components$': 'E:\\projects\\glamping\\backend\\node_modules\\styled-components',
      'yup$': 'E:\\projects\\glamping\\backend\\node_modules\\yup'
    },
    extensions: [ '.js', '.jsx', '.react.js', '.ts', '.tsx' ]
  },
  module: {
    rules: [
      [Object], [Object],
      [Object], [Object],
      [Object], [Object],
      [Object]
    ]
  },
  plugins: [
    HtmlWebpackPlugin { userOptions: [Object], version: 5 },
    DefinePlugin { definitions: [Object] },
    MiniCssExtractPlugin {
      _sortedModulesCache: [WeakMap],
      options: [Object],
      runtimeOptions: [Object]
    }
  ],
  stats: 'errors-only',
  mode: 'production',
  bail: true,
  devtool: false,
  output: {
    path: 'E:\\projects\\glamping\\backend\\build',
    publicPath: '/admin/',
    filename: '[name].[contenthash:8].js',
    chunkFilename: '[name].[contenthash:8].chunk.js'
  },
  optimization: {
    minimize: true,
    minimizer: [ [ESBuildMinifyPlugin] ],
    moduleIds: 'deterministic',
    runtimeChunk: true
  }
}
/ Building admin panel[ERROR] Module not found: Error: Can't resolve '....srcpluginspayment-gateway/strapi-admin' in 'E:\projects\glamping\backend\.strapi\client'
× Building admin panel
[ERROR]  There seems to be an unexpected error, try again with --debug for more information

[ERROR] Command failed with exit code 1: C:\Program Files\nodejs\node.exe E:\projects\glamping\backend\node_modules\@strapi\strapi\bin\strapi.js build --debug
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
PS E:\projects\glamping\backend> 

`

I have the same problem when adding a custom plugin in typescript. Error message is 'Development/open-course/backend/src/plugins/course/strapi-server.js: Cannot find module './dist/server'`

--- a/docs/docs/docs/01-core/admin/06-commands/01-build.md
+++ b/docs/docs/docs/01-core/admin/06-commands/01-build.md
@@ -45,6 +45,10 @@ interface BuildContext {
    */
   basePath: string;
   /**
+   * The customisations defined by the user in their app.js file
+   */
+  customisations?: AppFile;
+  /**
    * The current working directory
    */
   cwd: string;
@@ -64,9 +68,9 @@ interface BuildContext {
    * The environment variables to be included in the JS bundle
    */
   env: Record<string, string>;
-  logger: Logger;
+  logger: CLIContext['logger'];
   /**
-   * The build or develop options
+   * The build options
    */
   options: Pick<BuildOptions, 'minify' | 'sourcemaps' | 'stats'> & Pick<DevelopOptions, 'open'>;
   /**
@@ -90,7 +94,7 @@ interface BuildContext {
    * The browserslist target either loaded from the user's workspace or falling back to the default
    */
   target: string[];
-  tsconfig?: TSConfig;
+  tsconfig?: CLIContext['tsconfig'];
 }
 ```
 

--- a/docs/docs/docs/01-core/admin/06-commands/02-develop.md
+++ b/docs/docs/docs/01-core/admin/06-commands/02-develop.md
@@ -24,7 +24,7 @@ Start your Strapi application in development mode
 Options:
   --polling         Watch for file changes (default: false) – Whether to use fs.watchFile (backed by polling), or fs.watch, this is passed directly to chokidar
   --no-build        [deprecated]: there is middleware for the server, it is no longer a separate process
-  --watch-admin     [deprecated]: there is now middleware for watching, it is no longer a separate process
+  --watch-admin     Watch the admin for changes (default: false)
   --browser <name>  [deprecated]: use open instead
   --open            Open the admin in your browser (default: true)
   -h, --help        Display help for command
@@ -66,6 +66,10 @@ interface DevelopOptions extends CLIContext {
    * The tsconfig to use for the build. If undefined, this is not a TS project.
    */
   tsconfig?: TsConfig;
+  /**
+   * Watch the admin for changes
+   */
+  watchAdmin?: boolean;
 }
 
 interface Logger {

--- a/examples/getstarted/src/admin/app.example.js
+++ b/examples/getstarted/src/admin/app.example.js
@@ -1,9 +1,0 @@
-const config = {
-  locales: ['fr'],
-};
-const bootstrap = () => {};
-
-export default {
-  config,
-  bootstrap,
-};

--- a/examples/getstarted/src/admin/app.js
+++ b/examples/getstarted/src/admin/app.js
@@ -1,0 +1,11 @@
+const config = {
+  locales: ['it', 'es', 'en'],
+};
+const bootstrap = () => {
+  console.log('I AM  BOOTSTRAPPED');
+};
+
+export default {
+  config,
+  bootstrap,
+};

--- a/packages/core/admin/_internal/node/core/admin-customisations.ts
+++ b/packages/core/admin/_internal/node/core/admin-customisations.ts
@@ -1,0 +1,32 @@
+import path from 'node:path';
+import { loadFile } from './files';
+
+const ADMIN_APP_FILES = ['app.js', 'app.mjs', 'app.ts', 'app.jsx', 'app.tsx'];
+
+interface AdminCustomisations {
+  config?: {
+    locales?: string[];
+  };
+  bootstrap?: Function;
+}
+
+interface AppFile {
+  path: string;
+  config: AdminCustomisations['config'];
+}
+
+const loadUserAppFile = async (appDir: string): Promise<AppFile | undefined> => {
+  for (const file of ADMIN_APP_FILES) {
+    const filePath = path.join(appDir, 'src', 'admin', file);
+    const configFile = await loadFile(filePath);
+
+    if (configFile) {
+      return { path: filePath, config: configFile };
+    }
+  }
+
+  return undefined;
+};
+
+export { loadUserAppFile };
+export type { AdminCustomisations, AppFile };

--- a/packages/core/admin/_internal/node/core/plugins.ts
+++ b/packages/core/admin/_internal/node/core/plugins.ts
@@ -1,9 +1,12 @@
 import os from 'node:os';
 import path from 'node:path';
+import fs from 'node:fs';
+import camelCase from 'lodash/camelCase';
 import { env } from '@strapi/utils';
 import { getModule, PackageJson } from './dependencies';
 import { loadFile } from './files';
-import { BuildContext, CreateBuildContextArgs } from '../createBuildContext';
+import { BuildContext } from '../createBuildContext';
+import { isError } from './errors';
 
 interface PluginMeta {
   name: string;
@@ -32,11 +35,11 @@ const validatePackageHasStrapi = (
 const validatePackageIsPlugin = (pkg: PackageJson): pkg is StrapiPlugin =>
   validatePackageHasStrapi(pkg) && pkg.strapi.kind === 'plugin';
 
-export const getEnabledPlugins = async ({
+const getEnabledPlugins = async ({
   strapi,
   cwd,
   logger,
-}: Pick<BuildContext, 'cwd' | 'logger' | 'strapi'>) => {
+}: Pick<BuildContext, 'cwd' | 'logger' | 'strapi'>): Promise<Record<string, PluginMeta>> => {
   const plugins: Record<string, PluginMeta> = {};
 
   /**
@@ -110,3 +113,73 @@ const loadUserPluginsFile = async (root: string): Promise<UserPluginConfigFile> 
 
   return {};
 };
+
+const getMapOfPluginsWithAdmin = (
+  plugins: Record<string, PluginMeta>,
+  { runtimeDir }: { runtimeDir: string }
+) =>
+  Object.values(plugins)
+    .filter((plugin) => {
+      if (!plugin) {
+        return false;
+      }
+
+      /**
+       * There are two ways a plugin should be imported, either it's local to the strapi app,
+       * or it's an actual npm module that's installed and resolved via node_modules.
+       *
+       * We first check if the plugin is local to the strapi app, using a regular `resolve` because
+       * the pathToPlugin will be relative i.e. `/Users/my-name/strapi-app/src/plugins/my-plugin`.
+       *
+       * If the file doesn't exist well then it's probably a node_module, so instead we use `require.resolve`
+       * which will resolve the path to the module in node_modules. If it fails with the specific code `MODULE_NOT_FOUND`
+       * then it doesn't have an admin part to the package.
+       */
+      try {
+        const isLocalPluginWithLegacyAdminFile = fs.existsSync(
+          //@ts-ignore
+          path.resolve(`${plugin.pathToPlugin}/strapi-admin.js`)
+        );
+
+        if (!isLocalPluginWithLegacyAdminFile) {
+          //@ts-ignore
+          let pathToPlugin = plugin.pathToPlugin;
+
+          if (process.platform === 'win32') {
+            pathToPlugin = pathToPlugin.split(path.sep).join(path.posix.sep);
+          }
+
+          const isModuleWithFE = require.resolve(`${pathToPlugin}/strapi-admin`);
+
+          return isModuleWithFE;
+        }
+
+        return isLocalPluginWithLegacyAdminFile;
+      } catch (err) {
+        if (isError(err) && 'code' in err && err.code === 'MODULE_NOT_FOUND') {
+          /**
+           * the plugin does not contain FE code, so we
+           * don't want to import it anyway
+           */
+          return false;
+        }
+
+        throw err;
+      }
+    })
+    .map((plugin) => {
+      const systemPath = plugin.isLocal
+        ? path.relative(runtimeDir, plugin.pathToPlugin)
+        : undefined;
+      const modulePath = systemPath ? systemPath.split(path.sep).join('/') : undefined;
+
+      return {
+        path: !plugin.isLocal
+          ? `${plugin.pathToPlugin}/strapi-admin`
+          : `${modulePath}/strapi-admin`,
+        name: plugin.name,
+        importName: camelCase(plugin.name),
+      };
+    });
+
+export { getEnabledPlugins, getMapOfPluginsWithAdmin };

--- a/packages/core/admin/_internal/node/core/plugins.ts
+++ b/packages/core/admin/_internal/node/core/plugins.ts
@@ -169,7 +169,7 @@ const getMapOfPluginsWithAdmin = (
     })
     .map((plugin) => {
       const systemPath = plugin.isLocal
-        ? path.relative(runtimeDir, plugin.pathToPlugin)
+        ? path.relative(runtimeDir, plugin.pathToPlugin.split('/').join(path.sep))
         : undefined;
       const modulePath = systemPath ? systemPath.split(path.sep).join('/') : undefined;
 

--- a/packages/core/admin/_internal/node/createBuildContext.ts
+++ b/packages/core/admin/_internal/node/createBuildContext.ts
@@ -14,6 +14,7 @@ import type { BuildOptions } from './build';
 import { DevelopOptions } from './develop';
 import { getEnabledPlugins } from './core/plugins';
 import { Strapi } from '@strapi/types';
+import { AppFile, loadUserAppFile } from './core/admin-customisations';
 
 interface BuildContext {
   /**
@@ -25,6 +26,10 @@ interface BuildContext {
    * this path so all asset paths will be rewritten accordingly
    */
   basePath: string;
+  /**
+   * The customisations defined by the user in their app.js file
+   */
+  customisations?: AppFile;
   /**
    * The current working directory
    */
@@ -166,9 +171,12 @@ const createBuildContext = async ({
 
   const target = browserslist.loadConfig({ path: cwd }) ?? DEFAULT_BROWSERSLIST;
 
+  const customisations = await loadUserAppFile(strapiInstance.dirs.app.root);
+
   const buildContext = {
     appDir: strapiInstance.dirs.app.root,
     basePath: `${adminPath}/`,
+    customisations,
     cwd,
     distDir,
     distPath,

--- a/packages/core/admin/_internal/node/createBuildContext.ts
+++ b/packages/core/admin/_internal/node/createBuildContext.ts
@@ -1,14 +1,11 @@
 import os from 'node:os';
 import path from 'node:path';
 import fs from 'node:fs/promises';
-import syncFs from 'node:fs';
-import camelCase from 'lodash/camelCase';
 import browserslist from 'browserslist';
 import strapiFactory, { CLIContext } from '@strapi/strapi';
 import { getConfigUrls } from '@strapi/utils';
 
 import { getStrapiAdminEnvVars, loadEnv } from './core/env';
-import { isError } from './core/errors';
 
 import type { BuildOptions } from './build';
 import { DevelopOptions } from './develop';

--- a/packages/core/admin/_internal/node/createBuildContext.ts
+++ b/packages/core/admin/_internal/node/createBuildContext.ts
@@ -12,7 +12,7 @@ import { isError } from './core/errors';
 
 import type { BuildOptions } from './build';
 import { DevelopOptions } from './develop';
-import { getEnabledPlugins } from './core/plugins';
+import { getEnabledPlugins, getMapOfPluginsWithAdmin } from './core/plugins';
 import { Strapi } from '@strapi/types';
 import { AppFile, loadUserAppFile } from './core/admin-customisations';
 
@@ -157,15 +157,7 @@ const createBuildContext = async ({
 
   logger.debug('Enabled plugins', os.EOL, plugins);
 
-  const pluginsWithFront = Object.values(plugins)
-    .filter(filterPluginsByAdminEntry)
-    .map((plugin) => ({
-      path: !plugin.isLocal
-        ? `${plugin.pathToPlugin}/strapi-admin`
-        : `${path.relative(runtimeDir, plugin.pathToPlugin)}/strapi-admin`,
-      name: plugin.name,
-      importName: camelCase(plugin.name),
-    }));
+  const pluginsWithFront = getMapOfPluginsWithAdmin(plugins, { runtimeDir });
 
   logger.debug('Enabled plugins with FE', os.EOL, plugins);
 
@@ -192,59 +184,6 @@ const createBuildContext = async ({
   } satisfies BuildContext;
 
   return buildContext;
-};
-
-interface Plugin extends Required<{}> {
-  name: string;
-}
-
-const filterPluginsByAdminEntry = (plugin: Plugin) => {
-  if (!plugin) {
-    return false;
-  }
-
-  /**
-   * There are two ways a plugin should be imported, either it's local to the strapi app,
-   * or it's an actual npm module that's installed and resolved via node_modules.
-   *
-   * We first check if the plugin is local to the strapi app, using a regular `resolve` because
-   * the pathToPlugin will be relative i.e. `/Users/my-name/strapi-app/src/plugins/my-plugin`.
-   *
-   * If the file doesn't exist well then it's probably a node_module, so instead we use `require.resolve`
-   * which will resolve the path to the module in node_modules. If it fails with the specific code `MODULE_NOT_FOUND`
-   * then it doesn't have an admin part to the package.
-   */
-  try {
-    const isLocalPluginWithLegacyAdminFile = syncFs.existsSync(
-      //@ts-ignore
-      path.resolve(`${plugin.pathToPlugin}/strapi-admin.js`)
-    );
-
-    if (!isLocalPluginWithLegacyAdminFile) {
-      //@ts-ignore
-      let pathToPlugin = plugin.pathToPlugin;
-
-      if (process.platform === 'win32') {
-        pathToPlugin = pathToPlugin.split(path.sep).join(path.posix.sep);
-      }
-
-      const isModuleWithFE = require.resolve(`${pathToPlugin}/strapi-admin`);
-
-      return isModuleWithFE;
-    }
-
-    return isLocalPluginWithLegacyAdminFile;
-  } catch (err) {
-    if (isError(err) && 'code' in err && err.code === 'MODULE_NOT_FOUND') {
-      /**
-       * the plugin does not contain FE code, so we
-       * don't want to import it anyway
-       */
-      return false;
-    }
-
-    throw err;
-  }
 };
 
 export { createBuildContext };

--- a/packages/core/admin/_internal/node/staticFiles.ts
+++ b/packages/core/admin/_internal/node/staticFiles.ts
@@ -8,16 +8,12 @@ import { DefaultDocument as Document } from '../../admin/src/components/DefaultD
 
 import type { BuildContext } from './createBuildContext';
 
-interface EntryModuleArgs {
-  plugins: BuildContext['plugins'];
-}
-
-const getEntryModule = ({ plugins }: EntryModuleArgs): string => {
-  const pluginsObject = plugins
+const getEntryModule = (ctx: BuildContext): string => {
+  const pluginsObject = ctx.plugins
     .map(({ name, importName }) => `'${name}': ${importName}`)
     .join(',\n');
 
-  const pluginsImport = plugins
+  const pluginsImport = ctx.plugins
     .map(({ importName, path }) => `import ${importName} from '${path}';`)
     .join('\n');
 
@@ -29,9 +25,19 @@ const getEntryModule = ({ plugins }: EntryModuleArgs): string => {
         ${pluginsImport}
         import { renderAdmin } from "@strapi/strapi/admin"
         
+        ${
+          ctx.customisations?.path
+            ? `import customisations from '${path.relative(
+                ctx.runtimeDir,
+                ctx.customisations.path
+              )}'`
+            : ''
+        }
+        
         renderAdmin(
           document.getElementById("strapi"),
           {
+            ${ctx.customisations?.path ? 'customisations,' : ''}
             plugins: {
         ${pluginsObject}
             }
@@ -88,7 +94,7 @@ const writeStaticClientFiles = async (ctx: BuildContext) => {
   ctx.logger.debug('Wrote the index.html file');
   await fs.writeFile(
     path.join(ctx.runtimeDir, 'app.js'),
-    format(getEntryModule({ plugins: ctx.plugins }), {
+    format(getEntryModule(ctx), {
       parser: 'babel',
     })
   );

--- a/packages/core/admin/_internal/node/webpack/config.ts
+++ b/packages/core/admin/_internal/node/webpack/config.ts
@@ -218,8 +218,19 @@ const mergeConfigWithUserConfig = async (config: Configuration, ctx: BuildContex
   const userConfig = await getUserConfig(ctx);
 
   if (userConfig) {
-    const webpack = await import('webpack');
-    return userConfig(config, webpack);
+    if (typeof userConfig === 'function') {
+      const webpack = await import('webpack');
+      return userConfig(config, webpack);
+    } else {
+      ctx.logger.warn(
+        `You've exported something other than a function from ${path.join(
+          ctx.appDir,
+          'src',
+          'admin',
+          'webpack.config'
+        )}, this will ignored.`
+      );
+    }
   }
 
   return config;

--- a/packages/core/admin/admin/src/render.ts
+++ b/packages/core/admin/admin/src/render.ts
@@ -5,10 +5,14 @@ import { createRoot } from 'react-dom/client';
 import { StrapiApp, StrapiAppConstructorArgs } from './StrapiApp';
 
 interface RenderAdminArgs {
+  customisations: StrapiAppConstructorArgs['adminConfig'];
   plugins: StrapiAppConstructorArgs['appPlugins'];
 }
 
-const renderAdmin = async (mountNode: HTMLElement | null, { plugins }: RenderAdminArgs) => {
+const renderAdmin = async (
+  mountNode: HTMLElement | null,
+  { plugins, customisations }: RenderAdminArgs
+) => {
   if (!mountNode) {
     throw new Error('[@strapi/admin]: Could not find the root element to mount the admin app');
   }
@@ -73,7 +77,7 @@ const renderAdmin = async (mountNode: HTMLElement | null, { plugins }: RenderAdm
   }
 
   const app = new StrapiApp({
-    adminConfig: {},
+    adminConfig: customisations,
     appPlugins: plugins,
   });
 

--- a/packages/core/strapi/src/admin.ts
+++ b/packages/core/strapi/src/admin.ts
@@ -5,8 +5,9 @@ import email from '@strapi/plugin-email/strapi-admin';
 // @ts-expect-error – No types, yet.
 import upload from '@strapi/plugin-upload/strapi-admin';
 
-const render = (mountNode: HTMLElement | null, { plugins }: RenderAdminArgs) => {
+const render = (mountNode: HTMLElement | null, { plugins, ...restArgs }: RenderAdminArgs) => {
   return renderAdmin(mountNode, {
+    ...restArgs,
     plugins: {
       'content-type-builder': contentTypeBuilder,
       // @ts-expect-error – TODO: fix this

